src/PermissionTree.js

```diff
@@ -0,0 +1,167 @@
+                    Object.values(methods).forEach((method) => {
+                        defaultTree[model.modelName][method.name] = {};
+
+                        this.ACCESS_TYPES.forEach((accessType) => {
+                            defaultTree[model.modelName][method.name][accessType] = false;
+                        });
+                    });
+                }
+            }
+        });
+
+        this.defaultTree = defaultTree;
+        return this.defaultTree;
+    }
+
+    getDefaultTree() {
+        if (this.defaultTree) {
+            return this.defaultTree;
+        }
+
+        return this._buildDefaultTree();
+    }
+
+
+    getUserPermissionTree(user) {
+        if (!user) {
+            throw new Error('No user was specified while trying to access the users permission tree!');
+        }
+
+        if (!this.userPermissionTrees.has(user.id)) {
+            // Deep clone
+            this.setUserPermissionTree(user, JSON.parse(JSON.stringify(this.getDefaultTree())));
+        }
+
+        return this.userPermissionTrees.get(user.id);
+    }
+
+    setUserPermissionTree(user, tree) {
+        this.userPermissionTrees.set(user.id, tree);
+    }
+
+
+    buildACLQueries(roleName) {
+        const promises = [];
+        const defaultTree = this.getDefaultTree();
+
+        Object.keys(defaultTree).forEach((modelName) => {
+            Object.keys(defaultTree[modelName]).forEach((remoteMethod) => {
+                Object.keys(defaultTree[modelName][remoteMethod]).forEach((accessType) => {
+                    promises.push(new Promise((resolve, reject) => {
+                        this.models.ACL.checkPermission(
+                            this.models.RoleMapping.ROLE,
+                            roleName,
+                            modelName,
+                            remoteMethod,
+                            accessType,
+                            (err, accessRequest) => {
+                                if (err) reject(err);
+
+                                resolve(accessRequest);
+                            },
+                        );
+                    }));
+                });
+            });
+        });
+
+
+        return promises;
+    }
+
+    async getACLPermissionsForRole(role) {
+        return Promise.all(this.buildACLQueries(role));
+    }
+
+    async createPermissionTree(user) {
+        if (!user || !user.userGroups) {
+            throw new Error('No user was specified while trying to load permission tree!');
+        }
+
+        // Note: Roles are passed as user groups (RoleMapping.GROUP)
+        await Promise.all(user.userGroups.map(async(group) => {
+            const accessRequests = await this.getACLPermissionsForRole(group);
+
+            this.updatePermissions(user, accessRequests);
+        }));
+    }
+
+    updatePermissions(user, accessRequests) {
+        accessRequests.forEach((accessRequest) => {
+            if (
+                !this.getPermission(
```

Comment:
Is there a reason you check first if the permission is not set to true yet? Is this important? Couldn't you just:
```Javascript
accessRequests.forEach((accessRequest) => {
   this.setPermission(user, accessRequest, !!accessRequest.isAllowed());
});
```

Reply:
This would not work IMO. Imagine the first role your user has grants you access and the second on doesn't so it would override the given access by the first gorup.